[Cmake] enable build of spdlog/fmt for all platforms

fuzzard opened the pull request:
## Description
Build from source spdlog and fmt for all platforms (except linux/freebsd unless specifically enabled of course)

## Motivation and context
Build from source spdlog and fmt for windows (and existing tools/depends builds)

## How has this been tested?
Testing various paths and combinations of existing libs of spdlog/fmt.

The particular corner case for all the extra stuff here is if spdlog lib exists (reports version 1.9.2), however fmt has been updated (eg 8.0.1 -> 8.1.0). We dont know what version of fmt spdlog was built against, so to make sure things are inline, allow fmt to set a property for the caller of the find_package(fmt) to know that fmt is being rebuilt. We can then choose to force a rebuild of spdlog to make sure its built against the updated fmt.

So when i bump fmt, it will force a rebuild of any existing spdlog lib to use the updated fmt.

@wsnipex this took me a bit to get my head around. Essentially this gives us the platform to rebuild a lib when its dependencies have been updated/rebuilt, whilst still being able to allow the quick path of using existing libs if they meet the version requirements.
EG we dont want to rebuild spdlog if its up to date and exists, and its dependencies havent changed.

@thexai if you find time to test this. The windows build of fmt and spdlog has been available for a while (#21108), so this just brings in the auto build for windows and try to reduce the requirement of rebuilds based on versions of libs

If anyone can come up with any other edge cases, do let me know. Im trying to cover everything i can within reason

## What is the effect on users?
N/A

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[cmake] default build spdlog/fmt internal all platforms

Commit:
[cmake] allow handling of more corner cases of existing lib versions of spdlog/fmt

If a dependency lib is rebuilt, allow a way to notify the calling parent that the dependency
is being built. This then allows the parent to rebuild to make sure all the correct
versions of a lib are inline.

Eg Spdlog exists and was built against FMT 7.0.1 (we cant know what version). We no longer
have FMT lib, so it is rebuilt (ver 8.0.1). Allow spdlog to recognise that its dependency
is being force built, and therefore it should rebuild to make sure the same lib/api is
used throughout.

Commit:
[cmake] modulehelper function to get only version from a specific VERSION file

The use of this is to allow a find module to get a version of a dependency
from a VERSION file without smashing the find module SETUP_BUILD_VARS variables to
allow us to provide a version to a find_package call of the dependency

thexai commented:
Tested it a few days ago and it worked fine (before rebase need). 

Logic to rebuild or reuse existent in `project\BuildDependencies\` tested too and seems right on Windows.

I'm working on Jenkins modifications necessary to isolate Matrix/Nexus depends and allow reuse previous.... (no force rebuild same dependences if not changed). 

Besides of trivial part of add exception to git clean command is need change build paths to no overwrite Matrix builds with Nexus, etc...

https://github.com/thexai/xbmc/commit/2fada7ec4cc15f4dff779343a7442d748f9141c1

In the end it will be more complicated than I thought before starting...  😅

garbear commented:
"We do these things not because they are easy, but because we thought they were going to be easy"

fuzzard committed:
[fmt] fixup patch for windows

Fixes patch applied for windows. Existing version doesnt install headers appropriately

Commit:
[cmake] dependent_option function

This function allows us to handle in a generic way our use of internal lib state,
and to then allow it to be overridden more easily by the user.

cmake_dependent_option has proven too restrictive for our use of it.

Commit:
[cmake] utilise new dependent_option function for internal deps

wsnipex reviewed (approved):
very nice
